tests/testthat/_snaps/dm.md

```diff
@@ -272,3 +272,78 @@
+      $ dst   <chr> "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A", "A",~
+      $ tzone <chr> "America/New_York", "America/New_York", "America/Chicago", "Amer~
+    Code
+      dm_nycflights13() %>% dm_zoom_to(flights) %>% glimpse(width = 100)
+    Output
+      dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
+      
+      Table: `flights`
+      4 outgoing foreign key(s):
+        (`carrier`) -> (`airlines$carrier`) no_action
+        (`origin`) -> (`airports$faa`) no_action
+        (`tailnum`) -> (`planes$tailnum`) no_action
+        (`origin`, `time_hour`) -> (`weather$origin`, `weather$time_hour`) no_action
+      Rows: 1,761
+      Columns: 19
+      $ year           <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 201~
+      $ month          <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ~
+      $ day            <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,~
+      $ dep_time       <int> 3, 16, 450, 520, 530, 531, 535, 546, 549, 550, 553, 553, 553, 553, 555, 555~
+      $ sched_dep_time <int> 2359, 2359, 500, 525, 530, 540, 540, 600, 600, 600, 600, 600, 600, 600, 600~
+      $ dep_delay      <dbl> 4, 17, -10, -5, 0, -9, -5, -14, -11, -10, -7, -7, -7, -7, -5, -10, -5, -4, ~
+      $ arr_time       <int> 426, 447, 634, 813, 824, 832, 1015, 645, 652, 649, 711, 837, 834, 733, 733,~
+      $ sched_arr_time <int> 437, 444, 648, 820, 829, 850, 1017, 709, 724, 703, 715, 910, 859, 759, 745,~
+      $ arr_delay      <dbl> -11, 3, -14, -7, -5, -18, -2, -24, -32, -14, -4, -33, -25, -26, -12, -19, 2~
+      $ carrier        <chr> "B6", "B6", "US", "UA", "UA", "AA", "B6", "B6", "EV", "US", "EV", "AA", "B6~
+      $ flight         <int> 727, 739, 1117, 1018, 404, 1141, 725, 380, 6055, 2114, 5716, 707, 507, 731,~
+      $ tailnum        <chr> "N571JB", "N564JB", "N171US", "N35204", "N815UA", "N5EAAA", "N784JB", "N337~
+      $ origin         <chr> "JFK", "JFK", "EWR", "EWR", "LGA", "JFK", "JFK", "EWR", "LGA", "LGA", "JFK"~
+      $ dest           <chr> "BQN", "PSE", "CLT", "IAH", "IAH", "MIA", "BQN", "BOS", "IAD", "BOS", "IAD"~
+      $ air_time       <dbl> 183, 191, 78, 215, 210, 149, 191, 39, 48, 36, 51, 201, 144, 85, 126, 94, 42~
+      $ distance       <dbl> 1576, 1617, 529, 1400, 1416, 1089, 1576, 200, 229, 184, 228, 1389, 1065, 50~
+      $ hour           <dbl> 23, 23, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6~
+      $ minute         <dbl> 59, 59, 0, 25, 30, 40, 40, 0, 0, 0, 0, 0, 0, 0, 0, 5, 0, 0, 0, 0, 0, 0, 0, ~
+      $ time_hour      <dttm> 2013-01-10 23:00:00, 2013-01-10 23:00:00, 2013-01-10 05:00:00, 2013-01-10 ~
+    Code
+      dm_nycflights13() %>% dm_zoom_to(weather) %>% select(-origin) %>% glimpse()
+    Output
+      dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
+      
+      Table: `weather`
+      Primary key: (`time_hour`)
```

Comment:
The pk should be removed here:

``` r
library(conflicted)
library(dm)

dm_nycflights13() %>%
  dm_select_tbl(weather) %>%
  dm_select(weather, -origin) %>%
  dm_get_all_pks()
#> # A tibble: 0 × 2
#> # … with 2 variables: table <chr>, pk_col <keys>
```

<sup>Created on 2022-06-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Thanks. Definitely a major issue to attend to.

Better?

``` r
library(dm)

dm_nycflights13() %>%
  dm_select_tbl(weather) %>%
  dm_select(weather, -origin) %>%
  glimpse()
#> dm of 1 tables: `weather`
#> 
#> Table: `weather`
#> Rows: 144
#> Columns: 14
#> $ year       <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013,…
#> $ month      <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
#> $ day        <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ hour       <int> 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 1…
#> $ temp       <dbl> 41.00, 39.02, 39.02, 39.92, 41.00, 41.00, 39.92, 41.00, 42.…
#> $ dewp       <dbl> 32.00, 30.02, 28.94, 26.96, 26.06, 26.06, 24.98, 24.98, 24.…
#> $ humid      <dbl> 70.08, 69.86, 66.85, 59.50, 54.97, 54.97, 54.81, 52.56, 48.…
#> $ wind_dir   <dbl> 230, 210, 230, 270, 320, 300, 280, 330, 330, 320, 320, 330,…
#> $ wind_speed <dbl> 8.05546, 9.20624, 6.90468, 5.75390, 6.90468, 12.65858, 6.90…
#> $ wind_gust  <dbl> NA, NA, NA, NA, NA, 20.71404, 17.26170, NA, NA, 26.46794, N…
#> $ precip     <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
#> $ pressure   <dbl> 1024.6, 1025.9, 1026.9, 1027.5, 1028.2, 1029.0, 1030.0, 103…
#> $ visib      <dbl> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ time_hour  <dttm> 2013-01-10 00:00:00, 2013-01-10 01:00:00, 2013-01-10 02:00…
```

<sup>Created on 2022-06-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Sorry to get involved again, but I think this was a misunderstanding: https://github.com/cynkra/dm/pull/1161/commits/77fba8ea93c43365950d03130f4b6ade955aa8de does not really change the behavior of the function, since for unzoomed dm's the result for `dm_get_pk_impl(x, table_name)` would be an empty list anyway.

I believe what Kirill meant was, that if one column that is part of a compound key -- be it PK or FK -- gets lost for a zoomed dm, we want to drop that key altogether, as it would be done in case the zoomed table would be updated or inserted.

``` r
library(dm)

# should return this:
dm_nycflights13() %>% dm_zoom_to(weather) %>% select(-origin) %>% glimpse()
#> dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
#> 
#> Table: `weather`
#> Rows: 144
#> Columns: 14
#> $ year       <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013,…
#> $ month      <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
#> $ day        <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ hour       <int> 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 1…
#> $ temp       <dbl> 41.00, 39.02, 39.02, 39.92, 41.00, 41.00, 39.92, 41.00, 42.…
#> $ dewp       <dbl> 32.00, 30.02, 28.94, 26.96, 26.06, 26.06, 24.98, 24.98, 24.…
#> $ humid      <dbl> 70.08, 69.86, 66.85, 59.50, 54.97, 54.97, 54.81, 52.56, 48.…
#> $ wind_dir   <dbl> 230, 210, 230, 270, 320, 300, 280, 330, 330, 320, 320, 330,…
#> $ wind_speed <dbl> 8.05546, 9.20624, 6.90468, 5.75390, 6.90468, 12.65858, 6.90…
#> $ wind_gust  <dbl> NA, NA, NA, NA, NA, 20.71404, 17.26170, NA, NA, 26.46794, N…
#> $ precip     <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
#> $ pressure   <dbl> 1024.6, 1025.9, 1026.9, 1027.5, 1028.2, 1029.0, 1030.0, 103…
#> $ visib      <dbl> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ time_hour  <dttm> 2013-01-10 00:00:00, 2013-01-10 01:00:00, 2013-01-10 02:00…

# and in case of renaming:
dm_nycflights13() %>% dm_zoom_to(weather) %>% rename(origin_location = origin) %>% glimpse()
#> dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
#> 
#> Table: `weather`
#> Primary key: (`origin_location`, `time_hour`)
#> Rows: 144
#> Columns: 15
#> $ origin_location <chr> "EWR", "EWR", "EWR", "EWR", "EWR", "EWR", "EWR", "EWR"…
#> $ year            <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, …
#> $ month           <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, …
#> $ day             <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10…
#> $ hour            <int> 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, …
#> $ temp            <dbl> 41.00, 39.02, 39.02, 39.92, 41.00, 41.00, 39.92, 41.00…
#> $ dewp            <dbl> 32.00, 30.02, 28.94, 26.96, 26.06, 26.06, 24.98, 24.98…
#> $ humid           <dbl> 70.08, 69.86, 66.85, 59.50, 54.97, 54.97, 54.81, 52.56…
#> $ wind_dir        <dbl> 230, 210, 230, 270, 320, 300, 280, 330, 330, 320, 320,…
#> $ wind_speed      <dbl> 8.05546, 9.20624, 6.90468, 5.75390, 6.90468, 12.65858,…
#> $ wind_gust       <dbl> NA, NA, NA, NA, NA, 20.71404, 17.26170, NA, NA, 26.467…
#> $ precip          <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, …
#> $ pressure        <dbl> 1024.6, 1025.9, 1026.9, 1027.5, 1028.2, 1029.0, 1030.0…
#> $ visib           <dbl> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10…
#> $ time_hour       <dttm> 2013-01-10 00:00:00, 2013-01-10 01:00:00, 2013-01-10 …
```

<sup>Created on 2022-06-27 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

And the same behavior for FKs.

For PKs you could achieve that for example with:
```
  # anticipate that some key columns could have been removed by the user
  if (is_zoomed(x)) {
    tracked_cols <- col_tracker_zoomed(x)
    pk <- if (all(pk[[1]] %in% tracked_cols)) {
      list(unname(prep_recode(tracked_cols)[pk[[1]]]))
    } else {
      NULL
    }
  }
```
but please feel free to change the implementation.

Reply:
Thanks for the clarification, Tobias! I was definitely confused about the expected output, so the example code really helped.

I've implemented this change for primary keys, but still need to handle the foreign keys.

Reply:
Is there an equivalent of `update_zoomed_pk()` for foreign keys?

Reply:
Good that you ask: at first I thought it is `update_zoomed_outgoing()` in R/zoom.R, but this function is called nowhere (and is also not giving the correct result).
The correct operation is done in `dm_insert_zoomed_outgoing_fks()`, in the calculation of the variable `new_out_keys`.
I would suggest to extract the code in this assignment and replace the unused function `update_zoomed_outgoing()` with it, if this makes sense.

Reply:
My apologies, actually `update_zoomed_outgoing()` is being called in `dm_update_zoomed()`. Maybe you can make use of it; but actually using the code in `dm_insert_zoomed_outgoing_fks()` still seems easiest to me. 
In any case I am not anymore of the opinion that `update_zoomed_outgoing()` should be replaced.